[sos_extras] A new plugin to collect user defined commands

pmoravec opened the pull request:
Enabled by presence of /etc/sos.extras.d/ directory, the plugin will
execute all commands from all files in the directory, in order to
collect extra user-defined data e.g. specific for a given system
or product (or not yet being collected by sosreport otherwise).

Every such file with commands must start by a header warning users to
consult with sos comunity the commands first, to prevent possibly
dangerous commands being executed.

Resolves: #1663

Signed-off-by: Pavel Moravec <pmoravec@redhat.com>

---
Please place an 'X' inside each '[]' to confirm you adhere to our [Contributor Guidelines](https://github.com/sosreport/sos/wiki/Contribution-Guidelines)

- [X] Is the commit message split over multiple lines and hard-wrapped at 72 characters?
- [X] Is the subject and message clear and concise?
- [X] Does the subject start with **[plugin_name]** if submitting a plugin patch or a **[section_name]** if part of the core sosreport code?
- [X] Does the commit contain a **Signed-off-by: First Lastname <email@example.com>**?

bmr-cymru commented:
As well as running commands isn't it useful to allow additional file system paths in this file to be collected? Maybe something like:

```
some_command_to_run --with --args
:/var/log/some_command.log -
:/var/log/some_command.big.log 0
```

Where lines beginning with `:` are a path to collect, with a size limit in the 2nd column?

I'm not sure I like the "magic disclaimer text" mechanism - I think it's good practice to explain what is and is not supported in this file, but I am not sure there's value in verifying that the disclaimer text is there before honouring the command list - this is a root-owned file in a path only writable by root. To some extent we have to trust that (and the presence of the text in the file we read doesn't really give any additional confidence). Perhaps it's better to add some warning text to the log / UI output when this plugin finds commands to execute in the file.

I think the current wording for the disclaimer can also be improved - right now it seems to imply that you /could/ put shell notation in this tile, but that you /shouldn't/ - in fact we just don't support shell notation at all (without an explicit `sh -c ...`).

pmoravec commented:
The extension to collecting files makes sense, I will implement it.

The disclaimer text: enforcing it as a head of any file with commands to execute sounds to me the only way how to ensure anyone playing with this feature *will* read the disclaimer, *before* using it. I am ok with any warning - e.g. just a warning printed to stdout whenever an extras file is processed? 

About the disclaimer text: that was just very first draft (of a content I am not fully sure of). What we shall warn about? Not using pipes or multiple commands ("cmd1; cmd2")? Not running commands on background? Ensure no command changes the system anyway (esp. removing a file or similar)?

And where the warning text should be? In manpages (not big audience), on web (and url link to it in the warning message)? Somewhere else?

pmoravec commented:
Updated the PR to collect files as well and without the disclaimer mandatory head.

pmoravec commented:
Good point wrt `--list-plugins`, fixed.

I also converted explicit `check_enabled` method to standard `files = ..` approach and rebased the PR to current upstream.
